### ISSUE TITLE
GPU Build process is broken

### ISSUE THREAD

**Issue by chancecardona** [SixK/coco-annotator-ng#4]
Hi, thank you for your contributions so far on this!

I am trying to build and test this and am noticing some things wrong with the GPU build.

First off, ./`build_base_image.sh` should be split out so that the commented lines are in another script, maybe `build_gpu_image.sh` containing these lines:
```
docker build . -t annotator_webclient --no-cache
docker build -f ./backend/workers/Dockerfile . -t annotator_workers --no-cache
```
Note how there is not `_gpu` suffix as this conflicts with the Dockerfile image names expected.

My real error is that when running `docker compose -f ./docker-compose.gpu.yaml up` (after running the previous scripts and build), I get this error with the `annotator_ia` container:
```
fatal: not a git repository (or any of the parent directories): .git
[2024-05-14 22:09:27 +0000] [7] [DEBUG] Current configuration:
  config: ia/gunicorn_config.py
  wsgi_app: None
  bind: ['0.0.0.0:6000']
  backlog: 2048
  workers: 1
  worker_class: eventlet
  threads: 1
  worker_connections: 1000
  max_requests: 0
  max_requests_jitter: 0
  timeout: 60
  graceful_timeout: 30
  keepalive: 2
  limit_request_line: 4094
  limit_request_fields: 100
  limit_request_field_size: 8190
  reload: True
  reload_engine: auto
  reload_extra_files: []
  spew: False
  check_config: False
  print_config: False
  preload_app: False
  sendfile: False
  reuse_port: False
  chdir: /workspace
  daemon: False
  raw_env: []
  pidfile: None
  worker_tmp_dir: None
  user: 0
  group: 0
  umask: 0
  initgroups: False
  tmp_upload_dir: None
  secure_scheme_headers: {'X-FORWARDED-PROTOCOL': 'ssl', 'X-FORWARDED-PROTO': 'https', 'X-FORWARDED-SSL': 'on'}
  forwarded_allow_ips: ['127.0.0.1']
  accesslog: None
  disable_redirect_access_to_syslog: False
  access_log_format: %(h)s %(l)s %(u)s %(t)s "%(r)s" %(s)s %(b)s "%(f)s" "%(a)s"
  errorlog: -
  loglevel: debug
  capture_output: False
  logger_class: gunicorn.glogging.Logger
  logconfig: None
  logconfig_dict: {}
  logconfig_json: None
  syslog_addr: udp://localhost:514
  syslog: False
  syslog_prefix: None
  syslog_facility: user
  enable_stdio_inheritance: False
  statsd_host: None
  dogstatsd_tags: 
  statsd_prefix: 
  proc_name: None
  default_proc_name: ia:app
  pythonpath: None
  paste: None
  on_starting: <function OnStarting.on_starting at 0x7a03575989d0>
  on_reload: <function OnReload.on_reload at 0x7a0357598af0>
  when_ready: <function WhenReady.when_ready at 0x7a0357598c10>
  pre_fork: <function Prefork.pre_fork at 0x7a0357598d30>
  post_fork: <function Postfork.post_fork at 0x7a0357598e50>
  post_worker_init: <function PostWorkerInit.post_worker_init at 0x7a0357598f70>
  worker_int: <function WorkerInt.worker_int at 0x7a0357599090>
  worker_abort: <function WorkerAbort.worker_abort at 0x7a03575991b0>
  pre_exec: <function PreExec.pre_exec at 0x7a03575992d0>
  pre_request: <function PreRequest.pre_request at 0x7a03575993f0>
  post_request: <function PostRequest.post_request at 0x7a0357599480>
  child_exit: <function ChildExit.child_exit at 0x7a03575995a0>
  worker_exit: <function WorkerExit.worker_exit at 0x7a03575996c0>
  nworkers_changed: <function NumWorkersChanged.nworkers_changed at 0x7a03575997e0>
  on_exit: <function OnExit.on_exit at 0x7a0357599900>
  ssl_context: <function NewSSLContext.ssl_context at 0x7a0357599a20>
  proxy_protocol: False
  proxy_allow_ips: ['127.0.0.1']
  keyfile: None
  certfile: None
  ssl_version: 2
  cert_reqs: 0
  ca_certs: None
  suppress_ragged_eofs: True
  do_handshake_on_connect: False
  ciphers: None
  raw_paste_global_conf: []
  strip_header_spaces: False
[2024-05-14 22:09:27 +0000] [7] [INFO] Starting gunicorn 21.2.0
[2024-05-14 22:09:27 +0000] [7] [DEBUG] Arbiter booted
[2024-05-14 22:09:27 +0000] [7] [INFO] Listening at: http://0.0.0.0:6000 (7)
[2024-05-14 22:09:27 +0000] [7] [INFO] Using worker: eventlet
[2024-05-14 22:09:27 +0000] [9] [INFO] Booting worker with pid: 9
[2024-05-14 22:09:27 +0000] [7] [DEBUG] 1 workers
/workspace/sam-hq/segment_anything/modeling/tiny_vit_sam.py:662: UserWarning: Overwriting tiny_vit_5m_224 in registry with segment_anything.modeling.tiny_vit_sam.tiny_vit_5m_224. This is because the name being registered conflicts with an existing name. Please check if this is not expected.
  return register_model(fn_wrapper)
/workspace/sam-hq/segment_anything/modeling/tiny_vit_sam.py:662: UserWarning: Overwriting tiny_vit_11m_224 in registry with segment_anything.modeling.tiny_vit_sam.tiny_vit_11m_224. This is because the name being registered conflicts with an existing name. Please check if this is not expected.
  return register_model(fn_wrapper)
/workspace/sam-hq/segment_anything/modeling/tiny_vit_sam.py:662: UserWarning: Overwriting tiny_vit_21m_224 in registry with segment_anything.modeling.tiny_vit_sam.tiny_vit_21m_224. This is because the name being registered conflicts with an existing name. Please check if this is not expected.
  return register_model(fn_wrapper)
/workspace/sam-hq/segment_anything/modeling/tiny_vit_sam.py:662: UserWarning: Overwriting tiny_vit_21m_384 in registry with segment_anything.modeling.tiny_vit_sam.tiny_vit_21m_384. This is because the name being registered conflicts with an existing name. Please check if this is not expected.
  return register_model(fn_wrapper)
/workspace/sam-hq/segment_anything/modeling/tiny_vit_sam.py:662: UserWarning: Overwriting tiny_vit_21m_512 in registry with segment_anything.modeling.tiny_vit_sam.tiny_vit_21m_512. This is because the name being registered conflicts with an existing name. Please check if this is not expected.
  return register_model(fn_wrapper)
[2024-05-14 22:09:29 +0000] [9] [ERROR] Exception in worker process
Traceback (most recent call last):
  File "/opt/conda/lib/python3.10/site-packages/gunicorn/arbiter.py", line 609, in spawn_worker
    worker.init_process()
  File "/opt/conda/lib/python3.10/site-packages/gunicorn/workers/geventlet.py", line 143, in init_process
    super().init_process()
  File "/opt/conda/lib/python3.10/site-packages/gunicorn/workers/base.py", line 134, in init_process
    self.load_wsgi()
  File "/opt/conda/lib/python3.10/site-packages/gunicorn/workers/base.py", line 146, in load_wsgi
    self.wsgi = self.app.wsgi()
  File "/opt/conda/lib/python3.10/site-packages/gunicorn/app/base.py", line 67, in wsgi
    self.callable = self.load()
  File "/opt/conda/lib/python3.10/site-packages/gunicorn/app/wsgiapp.py", line 58, in load
    return self.load_wsgiapp()
zz info: vit_b /models/sam_hq_vit_b.pth
  File "/opt/conda/lib/python3.10/site-packages/gunicorn/app/wsgiapp.py", line 48, in load_wsgiapp
    return util.import_app(self.app_uri)
  File "/opt/conda/lib/python3.10/site-packages/gunicorn/util.py", line 371, in import_app
    mod = importlib.import_module(module)
  File "/opt/conda/lib/python3.10/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/workspace/ia/__init__.py", line 14, in <module>
    from .api import blueprint as api
  File "/workspace/ia/api/__init__.py", line 6, in <module>
    from .models import api as ns_models
  File "/workspace/ia/api/models.py", line 20, in <module>
    from ..util.sam import model as sam
  File "/workspace/ia/util/sam.py", line 49, in <module>
    model = SAM()
  File "/workspace/ia/util/sam.py", line 19, in __init__
    sam = sam_model_registry[SAM_MODEL_TYPE](checkpoint=SAM_MODEL_PATH)
  File "/workspace/sam-hq/segment_anything/build_sam.py", line 38, in build_sam_vit_b
    return _build_sam(
  File "/workspace/sam-hq/segment_anything/build_sam.py", line 162, in _build_sam
    state_dict = torch.load(f, map_location=device)
  File "/opt/conda/lib/python3.10/site-packages/torch/serialization.py", line 795, in load
    return _legacy_load(opened_file, map_location, pickle_module, **pickle_load_args)
  File "/opt/conda/lib/python3.10/site-packages/torch/serialization.py", line 1002, in _legacy_load
    magic_number = pickle_module.load(f, **pickle_load_args)
_pickle.UnpicklingError: invalid load key, '<'.
[2024-05-14 22:09:29 +0000] [9] [INFO] Worker exiting (pid: 9)
[2024-05-14 22:09:29 +0000] [7] [ERROR] Worker (pid:9) exited with code 3
[2024-05-14 22:09:29 +0000] [7] [ERROR] Shutting down: Master
[2024-05-14 22:09:29 +0000] [7] [ERROR] Reason: Worker failed to boot.
```

Any ideas?

**Comment by SixK** [comment]
HI,

The right way to build is to first run:  
bash ./build_base_image.sh

Then:  
docker compose -f ./docker-compose.gpu.yml build

once done you can run with:  
docker compose -f ./docker-compose.gpu.yml up

Note that docker-compose.gpu.yml will probably not work (I've just seen that I can't login when trying to run it on my development sources.). But you should not have errors you have.
You can use ./docker-compose.dev.yml yaml file instead.
It should do the same things, with just some more debug ( interface is accessible by localhost:8080 in this case)

Commented lines in build_base_image.sh are old lines from original jsbroks project.
I have to clean them since now theses lines are useless.

I will try to find why I can't login anymore when running docker-compose.gpu.yml

**Comment by SixK** [comment]
OK, my bad ! :) 
./docker-compose.gpu.yml (and /docker-compose.yml) also works.
It just don't use the same database docker volume name than ./docker-compose.dev.yml  , so database is empty and contains no user by default. Everything seem's ok once a new user is created, just everything has to be recreated...

I may have a look to use the same database docker volume name for ./docker-compose.gpu.yml , ./docker-compose.yml and ./docker-compose.dev.yml

**Comment by SixK** [comment]
problem about docker-compose.gpu.yml not using the same docker volume than other docker-compose files should be fixed with commit 39a46f77f0605302ebc08e3387d26b0a39b9456e

**Comment by SixK** [comment]
Reading your comment once again let me think you have build docker images as it should.

The "_pickle.UnpicklingError: invalid load key, '<'. " error let me think you don't have downloaded sam-hq model first ?

Try this:
```
cd models
bash ./sam_hq_model.sh
```
You will probably also need to download model for DEXTR segmentation tool:
`bash ./dextr_pascal_sbd.sh`

By default, the sam-hd vit_b model is used, but if you have a good GPU with enough memory, you can use  more robust models like vit_l or vit_h. (or weaker vit_tiny also)
you will have to download this models first from official sam-hq github:
https://github.com/SysCV/sam-hq

and change the following lines in docker-compose.gpu.yml:
      - SAM_MODEL_FILE=/models/sam_hq_vit_b.pth
      - SAM_MODEL_TYPE=vit_b

**Comment by chancecardona** [comment]
Interesting, thanks for the updates. The first error I mentioned stands, namely that you don't define `annotator_webclient` image anywhere (other than in the commented out build commands, meaning that they are not optional).

Also, we won't need all those docker-compose.yaml's if you use `extends:` keyword when defining the services. You can have the first file define most things, then the gpu file just extend it with the GPU `capabilities`. 

I have manually downloaded the model weights you mentioned and pointed everything correctly but I still get the exact same error with the `annotator_ia`. Edit: However it seems to still work? Investigating deeper.

**Comment by chancecardona** [comment]
It seems that no data I put in a dataset directory ever gets populated to a dataset. I don't think I had this error in the non ng version

**Comment by SixK** [comment]
For dataset populating, it may be related to the fact I had to disable python watchdog module.
watchdog was freezing gunicorn/flask backend.
I have to try to reactivate it since watchdog is now v4.0.0 (instead of v3.0.0 in original coco annotator)

As a workaround, you can go in dataset page and click to scan button (at the left side under Generate). 
It should scan the dataset directory to find images.
Note that images may not appear instantly, you may have to refresh the page.
You can follow scan advance in tasks tab.

Your right for "extends:", I have to explore this option. 
I tried to keep the original coco-annotator logic.

For annotator_webclient image, you have to run:
`docker compose -f ./docker-compose.build.yml build`

Sorry, I kept original coco-annotator logic (I hope so) --> https://github.com/jsbroks/coco-annotator/blob/master/docker-compose.build.yml
I have to see If I can do something better.

For Sam-HQ, I delete my model file and downloaded it again using my script.
I had the same error.
I dunno if hugging-face moved the file to git-lfs or if I've put the wrong link, 
but link is actually wrong. (it get the http source page)
you can download the right file using: 
`wget https://huggingface.co/lkeab/hq-sam/resolve/main/sam_hq_vit_b.pth`

I have to fix it...

**Comment by SixK** [comment]
model address should be fixed by commit 3849f9b86e858d269c10722ca11e2b6967d683c4

**Comment by chancecardona** [comment]
Thank you so much! Tool now works beautifully with the vit_b model for me (other dockers work as well). A refresh fixed the dataset issue I had. If I get time before you do I will look into simplifying the build process also, but may require pushing an image to dockerhub so we can tackle in a later issue.
